HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,44 @@
+def BFS(sc, graph, name):
+
+    # Note: on the Virtual machine, things work best if everything is partitioned into 2.  
+    
+    # Rdd of name, shortest distance tuples that will be returned.  In each iteration, heros are added to this.  Initialize with the source node, and partition into 2 partitions to match the graph.
+    rdd = sc.parallelize([(name, 0)], 2).cache()
+    
+    # This RDD is used intermediately to build a collection of connections to build on
+    namesRdd = rdd
+
+    # Iterator used in the while loop
+    i = 0
+       
+    acc = sc.accumulator(0)    
+
+    while acc.value == 0:
+
+        i+=1
+
+        # Use join to look up the connections of the next round of connections
+        namesRdd = graph.join(namesRdd)
+
+        # Use mapValues to take just the list of connections and remove the distance
+        namesRdd = namesRdd.mapValues(lambda (x,y): x).partitionBy(2)
+
```

Comment:
I experimented with 4, 8, etc. and they were slower.  I think that the Virtual Machine doesn't have as many CPUs, so partitionBy(2) gave me my best result.